pkg/async/cond.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Cond.go provides a context respecting sync condition
+
+package async
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// Cond is a sync.Cond that respects context cancellation.
+// It provides equivalent functionality to sync.Cond (excepting there is no `Signal` method), except that
+// the Wait method exits with error if the context cancels.
+type Cond struct {
+	pointer atomic.Pointer[chan struct{}]
+}
+
+// ch returns the channel that Waiters are waiting on, possibly creating one if it doesn't exist.
+func (c *Cond) ch() chan struct{} {
+	t := make(chan struct{})
+	c.pointer.CompareAndSwap(nil, &t)
+	return *c.pointer.Load()
```

Comment:
This is going to create a lot of garbage objects if `ch()` is invoked in a hot path. Maybe optimize?
```suggestion
	value := c.pointer.Load()
	if value == nil {
		t := make(chan struct{})
		c.pointer.CompareAndSwap(nil, &t)
		value = c.pointer.Load()
	}
	return *value
```

Reply:
good point, and good idea.

Reply:
i did something a little different, but similar